### PR TITLE
move libblockmanager to libblockchain && add pbftconsensus implementation && submit blockverifier interface

### PR DESCRIPTION
**Pull request by cyjseagull**
1. move libblockmanager to libblockchain && modify related files correspondingly(include libconsensus, libtxpool and unit test for libtxpool)
2. add pbftconsensus implementation
3. submit blockverifier interface
4. submit other files related to blockverifier(sync from https://github.com/mingzhenliu/lab-bcos/tree/dev)

### CODE REVIEW

File: libconsensus/pbft/PBFTConsensus.cpp [hunk 1]
```diff
@@ -0,0 +1,903 @@
...
+        /// peer is a miner ?
+        if (getIndexByMiner(session.nodeID) < 0)
+            return;
+        /// peer is in the _filter list
+        if (filter.count(session.nodeID))
+        {
+            broadcastMark(session.nodeID, packetType, key);
+            return;
+        }
+        if (broadcastFilter(session.nodeID, packetType, key))
```
Comment (wheatli) [inline libconsensus/pbft/PBFTConsensus.cpp]: plz double check the logic

File: libconsensus/pbft/PBFTConsensus.cpp [hunk 2]
```diff
@@ -0,0 +1,903 @@
...
+#endif
+        LOG(ERROR) << "Wrong Miner List, right_miner_size = " << m_minerList.size()
+                   << " miner size of block_hash = " << block.blockHeader().hash() << " is "
+                   << block.blockHeader().sealerList().size();
+        BOOST_THROW_EXCEPTION(
+            BlockMinerListWrong() << errinfo_comment("Wrong Miner List of Block"));
+    }
+}
+
+Sealing PBFTConsensus::execBlock(PrepareReq& req, ostringstream& oss)
```
Comment (wheatli) [inline libconsensus/pbft/PBFTConsensus.cpp]: return Sealing is expensive

File: libconsensus/pbft/PBFTConsensus.cpp [hunk 3]
```diff
@@ -0,0 +1,903 @@
...
+        {
+            LOG(INFO) << "Discard this block, blk_no=" << m_reqCache->prepareCache().height
+                      << ",highest_block=" << m_highestBlock.number();
+        }
+    }
+}
+
+void PBFTConsensus::reportBlock(BlockHeader const& blockHeader)
+{
+    /// update the highest block
```
Comment (wheatli) [inline libconsensus/pbft/PBFTConsensus.cpp]: do you forget to protect these data by mutex? plz double check it

File: libconsensus/Common.h [hunk 4]
```diff
@@ -128,36 +242,114 @@ struct PBFTMsg
...
+    void updatePrepareReq(Sealing& sealing, KeyPair const& keyPair)
     {
-        RLP rlp(data);
-        populate(rlp);
+        timestamp = u256(utcTime());
+        block_hash = sealing.block.blockHeader().hash();
+        sig = signHash(block_hash, keyPair);
+        sig2 = signHash(fieldsWithoutBlock(), keyPair);
+        bytes block_data;
+        sealing.block.encode(block_data);
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/Common.h]: already modified, please refer to:  https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-80a0be1a017ddb493dc39f286f841732R156

File: libconsensus/pbft/PBFTReqCache.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    std::vector<std::pair<u256, Signature>> sig_list;
+    for (auto item : m_commitCache[m_prepareCache.block_hash])
+    {
+        sig_list.push_back(std::make_pair(item.second.idx, Signature(item.first.c_str())));
+    }
+    assert(u256(sig_list.size()) >= minSigSize);
+    /// set siglist for prepare cache
+    dev::eth::Block block(m_prepareCache.block);
+    block.setSigList(sig_list);
+    return block;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTReqCache.cpp]: already modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-8bf64877f6eaeebb0fff262a3fa80b09R47

File: libconsensus/pbft/PBFTReqCache.h [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    {
+        return cacheExists(m_commitCache, req.block_hash, req.sig.hex());
+    }
+
+    bool isExistViewChange(ViewChangeReq const& req)
+    {
+        return cacheExists(m_recvViewChangeReq, req.block_hash, req.idx);
+    }
+
+    u256 getSigCacheSize(h256 const& blockHash) { return u256(m_signCache[blockHash].size()); }
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTReqCache.h]: already modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-524ee7e94a901f6ba49290a21f30dba5R76

File: libconsensus/pbft/PBFTConsensus.h [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+    }
+    void clearMask() { m_broadCastCache->clearAll(); }
+    ssize_t getIndexByMiner(dev::h512 node_id) const;
+    h512 getMinerByIndex(size_t index) const;
+
+    /// trans data into message
+    Message::Ptr transDataToMessage(
+        bytesConstRef data, uint16_t const& packetType, uint16_t const& protocolId)
+    {
+        Message::Ptr message;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.h]: already modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-4e84504bf7fa2ce1c95ef09ed00e5cfeR241

File: libconsensus/pbft/PBFTConsensus.h [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+    bool m_emptyBlockFlag;
+    /// whether to omit empty block
+    bool m_omitEmptyBlock;
+    // backup msg
+    std::shared_ptr<dev::db::LevelDB> m_backupDB = nullptr;
+
+    /// get group id
+    uint8_t m_groupId;
+    /// static vars
+    const static std::string c_backupKeyCommitted;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.h]: please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-4e84504bf7fa2ce1c95ef09ed00e5cfeR413

File: libconsensus/pbft/PBFTConsensus.h [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+    template <class T>
+    bool hasConsensused(T const& req) const
+    {
+        return req.height < m_consensusBlockNumber || req.view < m_view;
+    }
+
+    template <typename T>
+    bool isFutureBlock(T const& req) const
+    {
+        return req.height > m_consensusBlockNumber || req.view > m_view;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.h]: already modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-4e84504bf7fa2ce1c95ef09ed00e5cfeR348

File: libconsensus/pbft/PBFTConsensus.h [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+    Message::Ptr transDataToMessage(bytesConstRef data, uint16_t const& packetType)
+    {
+        return transDataToMessage(data, packetType, m_protocolId);
+    }
+
+    bool isValidReq(Message::Ptr message, std::shared_ptr<Session> session)
+    {
+        h512 node_id;
+        bool is_miner = isMiner(node_id, m_idx);
+        if (message->buffer()->size() <= 0 || session->id() == node_id || !is_miner)
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.h]: already modified: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-4e84504bf7fa2ce1c95ef09ed00e5cfeR257

File: libconsensus/pbft/PBFTConsensus.cpp [hunk 11]
```diff
@@ -0,0 +1,903 @@
...
+void PBFTConsensus::broadcastSignReq(PrepareReq const& req)
+{
+    SignReq sign_req(req, m_keyPair, m_idx);
+    bytes sign_req_data;
+    sign_req.encode(sign_req_data);
+    broadcastMsg(SignReqPacket, sign_req.sig.hex(), ref(sign_req_data));
+    m_reqCache->addSignReq(sign_req);
+}
+
+inline bool PBFTConsensus::isMiner(h512& nodeID, const u256& idx) const
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.cpp]: already modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-7739c3316db232bb609da12413f4982eR286

File: libconsensus/pbft/PBFTConsensus.cpp [hunk 12]
```diff
@@ -0,0 +1,903 @@
...
+
+void PBFTConsensus::broadcastMsg(unsigned const& packetType, std::string const& key,
+    bytesConstRef data, std::unordered_set<h512> const& filter)
+{
+    auto sessions = m_service->sessionInfos();
+    for (auto session : sessions)
+    {
+        /// peer is a miner ?
+        if (getIndexByMiner(session.nodeID) < 0)
+            return;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.cpp]: logic has been modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-7739c3316db232bb609da12413f4982eR327

File: libconsensus/pbft/PBFTConsensus.cpp [hunk 13]
```diff
@@ -0,0 +1,903 @@
...
+        /// peer is in the _filter list
+        if (filter.count(session.nodeID))
+        {
+            broadcastMark(session.nodeID, packetType, key);
+            return;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.cpp]: logic has been modified, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-7739c3316db232bb609da12413f4982eR327

File: libconsensus/pbft/PBFTConsensus.cpp (hunk 3 above)
Comment (cyjseagull) [reply libconsensus/pbft/PBFTConsensus.cpp]: mutex has been added, please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/83/files#diff-7739c3316db232bb609da12413f4982eR587

**Pull request merged**